[JJtan2002] iP

JJtan2002 opened the pull request:
# iP PR Description Update

This PR introduces several **improvements** and _bug fixes_ to the iP project. Below are the main changes:

## Key Changes
- **Bug Fixes:**
  - 🐞 Fixed a bug that caused the `Parser` class to incorrectly handle invalid commands.
- **Enhancements:**
  - Improved error handling in the `EchoBot` to better guide the user on invalid input.
- **Refactoring:**
  - Refactored code to apply more Object-Oriented (OO) principles.

## Steps to Test the Changes
1. Clone the repository: 
    ```bash
    git clone https://github.com/your-repo/iP.git
    ```
2. Navigate to the project directory:
    ```bash
    cd iP
    ```
3. Run the application:
    ```bash
    ./run.sh
    ```
4. Test the new functionalities using the sample inputs provided in the `README.md`.

## Tasks Completed
- [x] Implemented basic command parsing.
- [x] Refactored the `Parser` class.
- [x] Updated documentation to reflect the latest changes.
- [ ] Add unit tests for new functionalities.

> “The details are not the details. They make the design.” – Charles Eames

## Additional Notes
- This PR addresses issue #1  and should improve the overall user experience.
- Refer to the [project guidelines](https://oss-generic.github.io/process/) for contributing to future enhancements.

## Code Snippet
Here's a key section of the refactored `Parser` class:

```java
// main method in EchoBot
    public static void main(String[] args) {
        // Load tasks when starting
        Storage.loadTasksFromFile(tasks);

        // Initialize parser and UI
        Parser parser = new Parser(tasks);
        Ui ui = new Ui();

        // Start the user interface
        ui.start();

        // Process user commands
        while (true) {
            String userInput = ui.nextInput();
            Command command = parser.parse(userInput);
            command.run();
            if (command.isExit()) {
                break;
            }
        }

        // Exit the user interface
        ui.exit();
    }

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

JJtan2002 committed:
Initial skeletal version of chatbot that simply greets the user and exits

JJtan2002 committed:
Allow input of commands

JJtan2002 committed:
Add storing and displaying text

JJtan2002 committed:
Added marking tasks as done

JJtan2002 committed:
Add support for tracking three types of tasks

JJtan2002 committed:
Added test for Echobot

JJtan2002 committed:
Deal with errors such as incorrect inputs entered by the user

JJtan2002 committed:
Add delete function

JJtan2002 committed:
Save and load tasks in hard disk automatically

JJtan2002 committed:
Merge branch-level-7 into master

JJtan2002 committed:
Use java LocalDate objects

JJtan2002 committed:
Merge pull request #1 from JJtan2002/branch-level-8

Use java LocalDate objects

JJtan2002 committed:
Initial refactoring of storage, tasklist

JJtan2002 committed:
Saving progress

JJtan2002 committed:
Refactor code to use more OOP

JJtan2002 committed:
Merge pull request #2 from JJtan2002/A-MoreOOP

A-MoreOOP

JJtan2002 committed:
Add packages

JJtan2002 committed:
Merge pull request #3 from JJtan2002/A-Packages

A-Packages

JJtan2002 committed:
Merge remote-tracking branch 'origin/add-gradle-support' into A-Gradle

JJtan2002 committed:
Merge pull request #4 from JJtan2002/A-Gradle

A-Gradle

JJtan2002 committed:
Add JUnit testing for Storage and Command

JJtan2002 committed:
Merge pull request #5 from JJtan2002/A-JUnit

A-JUnit

JJtan2002 committed:
Package the app as an executable JAR file so that it can be distributed easily.

JJtan2002 committed:
Merge pull request #6 from JJtan2002/A-Jar

A-Jar

JJtan2002 committed:
Add command to find a task by searching for keyword

JJtan2002 committed:
Merge pull request #7 from JJtan2002/branch-level-9

Level 9. Find

JJtan2002 committed:
Added JavaDoc comment headers

JJtan2002 committed:
Merge pull request #8 from JJtan2002/branch-level-9

Added JavaDoc comment headers

JJtan2002 committed:
Merge pull request #9 from JJtan2002/A-JavaDoc

A-JavaDoc

JJtan2002 committed:
Follow coding standards

JJtan2002 committed:
Merge pull request #10 from JJtan2002/A-CodingStandard

Add coding standards

calebchongsj commented on src/main/java/echobot/Command.java:
```diff
@@ -0,0 +1,367 @@
+     */
+    public void run() {
+        switch (this.command) {
+            case "bye":
```
Kindly indent your switch code. For instance, the 'case' keyword should be indented towards the left inside the switch block of code.

calebchongsj commented on src/main/java/echobot/Command.java:
```diff
@@ -0,0 +1,367 @@
+                    System.out.println("____________________________________________________________");
+                }
+            }
+            catch (NumberFormatException e) {
```
Indent your catch line to be on the same line as the closing braces of the try block.

calebchongsj commented on src/main/java/echobot/Command.java:
```diff
@@ -0,0 +1,367 @@
+package echobot;
+import echobot.task.*;
```
Do not import using the * notation. Imported classes should always be listed explicitly.

calebchongsj reviewed (commented):
Looks good to merge. Just some minor changes to fix. Great job!

JJtan2002 committed:
Ensure code adheres to style standards

JJtan2002 committed:
Merge pull request #11 from JJtan2002/A-CheckStyle

Ensure code adheres to style standards

JJtan2002 committed:
Add JavaFX GUI

JJtan2002 committed:
Merge pull request #12 from JJtan2002/Level-10

Add JavaFX GUI

JiaYi-Gallium369 commented on src/main/java/echobot/EchoBot.java:
```diff
@@ -0,0 +1,77 @@
+import echobot.task.TaskList;
+
+import java.io.ByteArrayOutputStream;
+import java.io.PrintStream;
```
standard library imports (`java.io.*`) should generally come before other imports. (`echobot`)

JiaYi-Gallium369 commented on src/main/java/echobot/Storage.java:
```diff
@@ -0,0 +1,126 @@
+ * Ensures that the file and directory exist and provides methods for saving and loading tasks.
+ */
+public class Storage {
+    private static final String FILE_PATH = "./data/echo-bot.txt";
```
Good job for setting the constant Strings as `final`!

JiaYi-Gallium369 commented on src/main/java/echobot/Storage.java:
```diff
@@ -0,0 +1,126 @@
+                    System.out.println(e.getMessage());
+                    // Optionally, delete or reset the file if corruption detected:
+                    // new File(FILE_PATH).delete(); // Resets the file
+                    // System.out.println("File reset due to corruption.");
```
can consider deleting these comments? as they should already be captured in your git history

JiaYi-Gallium369 commented on src/main/java/echobot/Ui.java:
```diff
@@ -0,0 +1,43 @@
+     * Informs the user that the EchoBot is ready to accept commands.
+     */
+    public void start() {
+        System.out.println("____________________________________________________________");
```
Can consider setting "____________________________________________________________" as a String constant?

JiaYi-Gallium369 reviewed (commented):
Code looks good to me! Overall it's easy to read with little to no coding standard violations

JJtan2002 committed:
- Simplified logic by refactoring deeply nested conditionals
- Enhanced readability and maintainability of the code

JJtan2002 committed:
Refactor: Reduce nested conditionals to improve code quality

- Simplified logic by refactoring deeply nested conditionals
- Enhanced readability and maintainability of the code

JJtan2002 committed:
Merge pull request #13

JJtan2002 committed:
Add assertions to Command class to ensure valid input and state

- Added assertions in the constructor to verify that `inputParts` is
  not null and contains at least one element.
- Ensured that `command` is not null or empty, and `tasks` is not
  null.
- Included assertions in the `run` method to check that `command` is
  not null before processing.

These changes help document assumptions and catch potential issues
during development by validating key conditions.

JJtan2002 committed:
Merge pull request #14

Add assertions to Command class to ensure valid input and state

JJtan2002 committed:
Create gradle.yml

JJtan2002 committed:
Provide in-App guidance to users
Add a command to access help page

JJtan2002 committed:
Merge pull request #15

Provide in-App guidance to users

JJtan2002 committed:
Added user guide

JJtan2002 committed:
Merge pull request #17

Add user guide

JJtan2002 committed:
Add GUI image to README

JJtan2002 committed:
Adjust build.gradle

JJtan2002 committed:
Fix coding style

JJtan2002 committed:
Fix code style

JJtan2002 committed:
Fix coding style

JJtan2002 committed:
Fix coding style

JJtan2002 committed:
Update gradle.yml
